File: app/models/pm/project.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


ProjectRole = Literal["ADMIN", "PROJECT_MANAGER", "DEVELOPER", "VIEWER"]

# 쓰기 가능한 역할 (Viewer 제외)
WRITABLE_ROLES: list[ProjectRole] = ["ADMIN", "PROJECT_MANAGER", "DEVELOPER"]


class ProjectCreate(BaseModel):
    org_id: str
    name: str = Field(..., min_length=1, max_length=100)
    key: str = Field(..., min_length=1, max_length=10, pattern=r"^[A-Z][A-Z0-9]*$")
    description: Optional[str] = Field(None, max_length=500)


class ProjectPatch(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)


class ProjectOut(BaseModel):
    id: str
    org_id: str
    name: str
    key: str
    description: Optional[str]
    is_sr_default: bool = False
    created_at: datetime
    updated_at: datetime


class ProjectMemberOut(BaseModel):
    id: str
    project_id: str
    user_id: str
    user_email: str
    user_name: str
    role: ProjectRole
    joined_at: datetime


class ProjectMemberAdd(BaseModel):
    user_id: str
    role: ProjectRole = "DEVELOPER"


class ProjectMemberRolePatch(BaseModel):
    role: ProjectRole

File: frontend/optool/src/services/pm/project.ts
import { api } from 'src/boot/axios'

export type ProjectRole = 'ADMIN' | 'PROJECT_MANAGER' | 'DEVELOPER' | 'VIEWER'

export type Project = {
  id: string
  orgId: string
  name: string
  key: string
  description: string | null
  isSrDefault: boolean
  createdAt: string
  updatedAt: string
}

export type ProjectMember = {
  id: string
  projectId: string
  userId: string
  userEmail: string
  userName: string
  role: ProjectRole
  joinedAt: string
}

export async function listProjects() {
  const { data } = await api.get<Project[]>('/pm/projects')
  return data
}

export async function createProject(payload: {
  org_id: string
  name: string
  key: string
  description?: string
}) {
  const { data } = await api.post<Project>('/pm/projects', payload)
  return data
}

export async function getProject(projectId: string) {
  const { data } = await api.get<Project>(`/pm/projects/${projectId}`)
  return data
}

export async function updateProject(projectId: string, payload: { name?: string; description?: string }) {
  const { data } = await api.patch<Project>(`/pm/projects/${projectId}`, payload)
  return data
}

export async function deleteProject(projectId: string) {
  await api.delete(`/pm/projects/${projectId}`)
}

export async function listProjectMembers(projectId: string) {
  const { data } = await api.get<ProjectMember[]>(`/pm/projects/${projectId}/members`)
  return data
}

export async function addProjectMember(projectId: string, userId: string, role: ProjectRole = 'DEVELOPER') {
  const { data } = await api.post<ProjectMember>(`/pm/projects/${projectId}/members`, { user_id: userId, role })
  return data
}

export async function changeProjectMemberRole(projectId: string, userId: string, role: ProjectRole) {
  const { data } = await api.patch<ProjectMember>(`/pm/projects/${projectId}/members/${userId}`, { role })
  return data
}

export async function removeProjectMember(projectId: string, userId: string) {
  await api.delete(`/pm/projects/${projectId}/members/${userId}`)
}

export async function setSrDefaultProject(projectId: string) {
  const { data } = await api.post<Project>(`/pm/projects/${projectId}/set-sr-default`)
  return data
}

export async function clearSrDefaultProject(projectId: string) {
  await api.delete(`/pm/projects/${projectId}/set-sr-default`)
}
